libs/knowledge-store/ragstack_knowledge_store/graph_store.py

```diff
@@ -326,52 +335,71 @@ def add_nodes(
         self,
         nodes: Iterable[Node] = None,
     ) -> Iterable[str]:
-        texts = []
-        metadatas = []
-        for node in nodes:
-            if not isinstance(node, TextNode):
-                raise ValueError("Only adding TextNode is supported at the moment")
-            texts.append(node.text)
-            metadatas.append(node.metadata)
-
-        text_embeddings = self._embedding.embed_texts(texts)
-
+        # Organize nodes by MIME type
+        mime_buckets = {}
         ids = []
 
-        tag_to_new_sources: Dict[str, List[Tuple[str, str]]] = {}
-        tag_to_new_targets: Dict[str, Dict[str, Tuple[str, List[float]]]] = {}
+        # Prepare nodes based on their type
+        for node in nodes:
+            if isinstance(node, TextNode):
+                if 'text' not in mime_buckets:
+                    mime_buckets['text'] = []
+                mime_buckets['text'].append(node)
+            if isinstance(node, Node) and node.mime_type:
+                main_mime_type = node.mime_type.split('/')[0]  # Split and take the first part, e.g., "image" from "image/png"
+                if main_mime_type not in mime_buckets:
+                    mime_buckets[main_mime_type] = []
+                mime_buckets[main_mime_type].append(node)
+            else:
+                raise ValueError("Unsupported node type")
+
+        # Process each MIME bucket
+        embeddings_dict = {}
+        for mime_type, nodes_list in mime_buckets.items():
+            method_name = f"embed_{mime_type}s"
+            if self._embedding.does_implement(method_name):
+                texts = [node.text if isinstance(node, TextNode) else node.content for node in nodes_list]
+                embeddings_dict[mime_type] = self._embedding.invoke(method_name, texts)
+            else:
+                # If no bulk method, try to call a singular method for each content
+                singular_method_name = f"embed_{mime_type}"
+                if self._embedding.does_implement(singular_method_name):
+                    embeddings = []
+                    for node in nodes_list:
+                        embedding = self._embedding.invoke(singular_method_name, node.text if isinstance(node, TextNode) else node.content)
+                        embeddings.append(embedding)
+                    embeddings_dict[mime_type] = embeddings
+                else:
+                    raise NotImplementedError(f"No embedding method available for MIME type: {mime_type}, implemented methods: {self._embedding.implements()}.")
+
 
         # Step 1: Add the nodes, collecting the tags and new sources / targets.
+        tag_to_new_sources = {}
```

Comment:
This has changed significantly from the previous implementation. I think it will need to be reworked if we work on adding this right now.